azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+        to be used.  (or a "copy data back" function needs to be added to this function
+        as an optional parameter.)
+        """
+
+        def translated_op_complete(op):
+            logger.info(
+                "%s: %s complete-returning result in %s",
+                self.name,
+                translated_op.name,
+                original_op.name,
+            )
+            original_op.error = translated_op.error
+            self.complete_op(original_op)
+
+        translated_op.callback = translated_op_complete
+        self.continue_op(translated_op)
+
+    # TODO: connected and disconnect migth become pipeline events.  maybe.
+    def on_connected(self):
+        """
+        Called by lower layers when the transport connects
+        """
+        if self.previous:
+            self.previous.on_connected()
+
+    def on_disconnected(self):
+        """
+        Called by lower layers when the transport disconnects
+        """
+        if self.previous:
+            self.previous.on_disconnected()
+
+
+class PipelineRoot(PipelineStage):
+    """
+    Object representing the root of a pipeline.  This is where the functions to build
+    the pipeline exist.  This is also where clients can add event handlers to receive
+    events from the pipeline.
+    """
+
+    def __init__(self):
+        PipelineStage.__init__(self)
+        self.on_pipeline_event = None
+
+    def _run_op(self, op):
+        """
+        run the operation.  At the root, the only thing to do is to pass the operation
+        to the next stage.
+        """
+        logger.info("{}({}) handling".format(self.name, op.name))
+        self.continue_op(op)
+
+    def add_next_stage(self, new_next_stage):
+        """
+        Add the next stage to the end of the pipeline.
+        """
+        old_tail = self
+        while old_tail.next:
+            old_tail = old_tail.next
+        old_tail.next = new_next_stage
+        new_next_stage.previous = old_tail
+        new_next_stage.pipeline_root = self
+        return self
+
+    def unhandled_error_handler(self, error):
+        """
+        Handler for errors that happen which cannot be tied to a specific operation.
+        This is still a tentative implimentation and masy be replaced by
+        some other mechanism as details on behavior are finalized.
+        """
+        # TODO: decide how to pass this error to the app
+        # TODO: if there's an error in the app handler, print it and exit
+        pass
+
+    def _handle_pipeline_event(self, event):
+        if self.on_pipeline_event:
+            # TODO: try/catch
+            self.on_pipeline_event(event)
+        else:
+            logger.warning("incoming pipeline event with no handler.  dropping.")
+
+
+class IotHub_UseSkAuthProvider(PipelineStage):
+    """
+    PipelineStage which handles a Shared Key Authentication Provider.  This is
+    likely a temporary stage if we're going to get rid of the two-stage
```

Comment:
Maybe? We are definitely getting rid of two-stage client creation, but I'm not sure we're getting rid of auth providers. They may just become internal.